desc/compute/utils.py

```diff
@@ -1347,3 +1360,776 @@ def body(i, mins):
     # The above implementation was benchmarked to be more efficient than
     # alternatives without explicit loops in GitHub pull request #501.
     return grid.expand(mins, surface_label)
+
+
+def polyint(c):
+    """Coefficients for the primitives of the given set of polynomials.
+
+    Parameters
+    ----------
+    c : ndarray
+        First axis should store coefficients of a polynomial.
+        For a polynomial given by ∑ᵢⁿ cᵢ xⁱ, where n is ``c.shape[0] - 1``,
+        coefficient cᵢ should be stored at ``c[n - i]``.
+
+    Returns
+    -------
+    poly : ndarray
+        Coefficients of polynomial primitive, ignoring the arbitrary constant.
+        That is, ``poly[i]`` stores the coefficient of the monomial xⁿ⁻ⁱ⁺¹,
+        where n is ``c.shape[0] - 1``.
+
+    """
+    poly = (c.T / jnp.arange(c.shape[0], 0, -1)).T
+    return poly
+
+
+def polyder(c):
+    """Coefficients for the derivatives of the given set of polynomials.
+
+    Parameters
+    ----------
+    c : ndarray
+        First axis should store coefficients of a polynomial.
+        For a polynomial given by ∑ᵢⁿ cᵢ xⁱ, where n is ``c.shape[0] - 1``,
+        coefficient cᵢ should be stored at ``c[n - i]``.
+
+    Returns
+    -------
+    poly : ndarray
+        Coefficients of polynomial derivative, ignoring the arbitrary constant.
+        That is, ``poly[i]`` stores the coefficient of the monomial xⁿ⁻ⁱ⁻¹,
+        where n is ``c.shape[0] - 1``.
+
+    """
+    poly = (c[:-1].T * jnp.arange(c.shape[0] - 1, 0, -1)).T
+    return poly
+
+
+def polyval(x, c):
+    """Evaluate the set of polynomials c at the points x.
+
+    Parameters
+    ----------
+    x : ndarray
+        Coordinates at which to evaluate the set of polynomials.
+        The first ``c.ndim`` axes should have shape ``c.shape[1:]``.
+    c : ndarray
+        First axis should store coefficients of a polynomial.
+        For a polynomial given by ∑ᵢⁿ cᵢ xⁱ, where n is ``c.shape[0] - 1``,
+        coefficient cᵢ should be stored at ``c[n - i]``.
+
+    Returns
+    -------
+    val : ndarray
+        ``val[j, k, ...]`` is the polynomial with coefficients ``c[:, j, k, ...]``
+        evaluated at the point ``x[j, k, ...]``.
+
+    Notes
+    -----
+    This function does not perform the same operation as
+    ``np.polynomial.polynomial.polyval(x, c)``.
+    An example usage of this function is shown in
+    tests/test_compute_utils.py::TestComputeUtils::test_polyval.
+
+    """
+    X = (x[jnp.newaxis].T ** jnp.arange(c.shape[0] - 1, -1, -1)).T
+    alphabet = "abcdefghijklmnopqrstuvwxyz"
+    sub = alphabet[: c.ndim]
+    val = jnp.einsum(f"{sub},{sub}...->{sub[1:]}...", c, X)
+    return val
+
+
+def tanh_sinh_quadrature(N, quad_limit=3.16):
```

Comment:
could set `quad_limit` based on the inverse tanh-sinh transform of 1-np.finfo.eps

Reply:
here's some related code from `quadax`: https://github.com/f0uriest/quadax/blob/640afd5bacc1a42f1e3c55c266fc5a02f668b2d1/quadax/utils.py#L166-L173

Reply:
Don't we just need to find the quad_limit once for a given precision? We are not using mixed precision. So it should be a fixed float.

Reply:
I think maybe the purpose is that machine epsilon is hardware dependent